Reorganize benchmark directory

aprokop opened the pull request:
This reorganization also allows to not instantiate all benchmarks for
every type.

This is driven by the desire to incorporate `BruteForce` from #468 into the benchmark testing with little trouble.

aprokop commented:
One item (not currently in the PR) needs discussion. By default, "all" option include Boost RTree and (in the future) could include BruteForce. As BruteForce will run for a long time with default parameters, a decision needs to be made. Either we reduce the sizes to run the test (from 50K to something like < 5K), or we will need to exclude some from the "all" option. In the same manner, it's unclear whether always running Rtree is worth it.

dalg24 commented on benchmarks/bvh_driver/spec.hpp:
```diff
@@ -0,0 +1,56 @@
+
+#include <string>
+
+#include <benchmark/benchmark.h>
```
Why did you include this header?

dalg24 commented on benchmarks/bvh_driver/bvh_driver.cpp:
```diff
@@ -477,6 +68,9 @@ class CmdLineArgs
   char **argv() { return _argv.data(); }
 };
 
+extern void register_bvh_benchmarks(Spec const &spec);
+extern void register_boostrtree_benchmarks(Spec const &spec);
```
Yuk

dalg24 replied:
I thought about it more and I think that is the very thing that I dislike about the way you structured the benchmark.
There is something messed up about the dependencies.  I would prefer a static vector to register functions that you loop through in that file (push back in the `benchmarks_foo.cpp` files).

That said I am not blocking.

dalg24 replied:
I shared https://github.com/dalg24/ArborX/commit/f332086bfeada3e0802ad50757954a075d733a10 on Slack

dalg24 commented on benchmarks/bvh_driver/benchmark_registration.hpp:
```diff
@@ -0,0 +1,394 @@
+#ifndef BENCHMARK_REGISTRATION_HPP
+#define BENCHMARK_REGISTRATION_HPP
+
+#include <ArborX_CrsGraphWrapper.hpp>
```
```suggestion
```

aprokop replied:
I'm not sure. I don't think we guarantee that `ArborX_CrsGraphWrapper.hpp` is included in `ArborX.hpp`.

dalg24 replied:
The `query()` free function for a given data structure is provided by its header.  This means all you need is include `<ArborX_LinearBVH.hpp>` or `<ArborX_BruteForce.hpp>`.

dalg24 commented on benchmarks/bvh_driver/benchmark_registration.hpp:
```diff
@@ -0,0 +1,394 @@
+#include <spec.hpp>
+
+template <typename DeviceType>
+inline Kokkos::View<ArborX::Point *, DeviceType>
```
```suggestion
Kokkos::View<ArborX::Point *, DeviceType>
```
and same below for `makeSpatialQueries` and `makeNearestQueries `

aprokop replied:
Don't think so. We do multiple tests instantiations for the same backend. For example, for Serial we would instantiate `makeSpatialQueries` and `makeNearestQueries` for boost `BVH<Serial>` and `BoostRTree`, resulting in linking failure.

masterleinad replied:
The point is that templates practically behave the same as `inline` functions even without the additional attribute. For me, this comes down to a matter of style and being consistent (where I don't lean too much in one direction or the other).

dalg24 replied:
There would be no link failure because these are template.  The `inline` specifier does not hurt but I objected to your adding it.  It is hard to tell what was simply moved from one file to another and what changed.  No need to inject superfluous changes in the mix.

aprokop replied:
While I don't understand how this works yet, I defer to you, and plan on reading through [stackoverflow](https://stackoverflow.com/questions/44335046/how-does-the-linker-handle-identical-template-instantiations-across-translation).

dalg24 commented on benchmarks/bvh_driver/benchmarks_bvh.cpp:
```diff
@@ -0,0 +1,92 @@
+  if (spec.backends == "all" || spec.backends == "cuda")
+    register_benchmark<Kokkos::Cuda,
+                       ArborX::BVH<typename Kokkos::Cuda::memory_space>>(
+        "ArborX::BVH<Cuda>", spec);
```
I had not noticed that before.  It is kinda sketchy.  Can be UVM depending on how Kokkos was built.

dalg24 replied:
Or was it the intent?

aprokop replied:
I'm not sure. This was introduced a long time ago, and I'm unsure of the intent at that time. Not addressing this in this PR.

dalg24 reviewed (commented):
I find the solution proposed overly complicated and I suspect you can achieve the same thing by splitting up the `register_benchmark` function that instantiate the benchmarks.

dalg24 commented:
> I find the solution proposed overly complicated and I suspect you can achieve the same thing by splitting up the `register_benchmark` function that instantiate the benchmarks.

See https://github.com/arborx/ArborX/pull/470#discussion_r566118031

masterleinad commented:
> I find the solution proposed overly complicated and I suspect you can achieve the same thing by splitting up the `register_benchmark` function that instantiate the benchmarks.

I agree here in that I would prefer to split `register_benchmark` to keep the changes minimal.

aprokop commented:
I completely rewrote the approach based on the feedback. Tell me if it's closer to what we want, I'm no longer sure I'm doing it right.

aprokop commented:
Any feedback on the new approach in this PR?

masterleinad commented:
> Any feedback on the new approach in this PR?

I'll have a look tomorrow.

masterleinad reviewed (commented):
Looks OK to me but do we still need the preprocessor defines?

aprokop commented on benchmarks/bvh_driver/benchmark_registration.hpp:
```diff
@@ -0,0 +1,374 @@
+  return queries;
+}
+
+#ifndef ARBORX_BENCHMARK_DISABLE_NEAREST_QUERY
```
You are right, we don't.

aprokop commented:
> Looks OK to me but do we still need the preprocessor defines?

Thanks, you are right, I missed removing these when working on the updated version. The should no longer be necessary.

aprokop commented:
@dalg24 I think I addressed all your concerns here by merging the spec header with the benchmark registration. Do you have anything else you wanted me to address?

dalg24 commented on benchmarks/bvh_driver/CMakeLists.txt:
```diff
@@ -9,7 +9,7 @@ find_package(Threads REQUIRED)
 add_executable(ArborX_BoundingVolumeHierarchy.exe bvh_driver.cpp)
 target_link_libraries(ArborX_BoundingVolumeHierarchy.exe ArborX::ArborX benchmark::benchmark Boost::program_options Threads::Threads)
-target_include_directories(ArborX_BoundingVolumeHierarchy.exe PRIVATE ${POINT_CLOUDS_INCLUDE_DIR} ${UNIT_TESTS_INCLUDE_DIR})
+target_include_directories(ArborX_BoundingVolumeHierarchy.exe PRIVATE ${POINT_CLOUDS_INCLUDE_DIR} ${UNIT_TESTS_INCLUDE_DIR} ${CMAKE_CURRENT_SOURCE_DIR})
```
I would use double quotes for headers in that directory instead but I am not blocking.

aprokop replied:
Followed the suggestion.

dalg24 reviewed (commented):
Looks good for the most part.  I'd prefer if you address my two comments but I am not blocking.

aprokop commented:
@dalg24 I addressed your latest comments.

dalg24 commented:
You forgot to drop the `${CMAKE_CURRENT_SOURCE_DIR}` in the include directories

aprokop commented:
> You forgot to drop the `${CMAKE_CURRENT_SOURCE_DIR}` in the include directories

According to the [reference](https://en.cppreference.com/w/cpp/preprocessor/include), the `#include "header"` form:
```
Searches for the file in implementation-defined manner. The intent of this syntax is to search for the files that are not controlled by the implementation. Typical implementations first search the directory where the current file resides and, only if the file is not found, search the standard include directories as with (1).
```
Thus, there is no guarantee that the current source directory is included (even though it is in most implementations).

aprokop commented:
I'm not sure why PGI distributed benchmark timed out, but it is not related to this PR.
